url8e6440c8aa_robdimsdale_project_euler_tgz-pJ8-com_rmd_personal_projecteuler_MainJ8/src/com/rmd/personal/projecteuler/p1to19/P015.java

```diff
@@ -0,0 +1,130 @@
+package com.rmd.personal.projecteuler.p1to19;
+
+import com.rmd.personal.projecteuler.Problem;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class P015 implements Problem {
+
+    private static final int DEFAULT_GRID_SIZE = 20;
+
+    private int gridSizeX;
+    private int gridSizeY;
+
+    private Map<XYPair, Long> routesAvailableMappedByXY;
+
+    public P015() {
+        this.setGridSizeX(DEFAULT_GRID_SIZE);
+        this.setGridSizeY(DEFAULT_GRID_SIZE);
+        routesAvailableMappedByXY = new HashMap<XYPair, Long>();
+    }
+
+    public int getGridSizeX() {
+        return gridSizeX;
+    }
+
+    public int getGridSizeY() {
+        return gridSizeY;
+    }
+
+    public void setGridSizeX(int gridSizeX) {
+        this.gridSizeX = gridSizeX;
+    }
+
+    public void setGridSizeY(int gridSizeY) {
+        this.gridSizeY = gridSizeY;
+    }
+
+    private Map<XYPair, Long> getRoutesAvailableMappedByXY() {
+        return routesAvailableMappedByXY;
+    }
+
+    @Override
+    public String getDescription() {
+        return "Starting in the top left corner of a 2×2 grid, and only being able to move to the right and down,\n"
+                + " there are exactly 6 routes to the bottom right corner.\n\n"
+                + "How many such routes are there through a 20×20 grid?";
+    }
+
+    @Override
+    public String run() {
+        long routesAvailable = this.numberOfRoutes(this.getGridSizeX(), this.getGridSizeY());
+        return String.valueOf(routesAvailable);
+    }
+
+    private long numberOfRoutes(int x, int y) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 78-88 into a separate method.